README
Test-Trap

Primarily (but not exclusively) for use in test scripts: A block eval
on steroids, configurable and extensible, but by default trapping
(Perl) STDOUT, STDERR, warnings, exceptions, would-be exit codes, and
return values from boxed blocks of test code.

INSTALLATION

To install this module, run the following commands:

  perl Makefile.PL
  make
  make test
  sudo make install

(If you are running as root, you may drop "sudo", and just run "make install".)

Alternatively, for those without a proper make, the Module::Build approach:

  perl Build.PL
  perl Build
  perl Build test
  sudo perl Build install

(Again, if you are running as root, you may run "make install" without "sudo".)


COPYRIGHT AND LICENCE

Copyright (C) 2006-2012 Eirik Berg Hanssen

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for Test-Trap

0.3.0	Thu Dec 18 21:57:51 CET 2014
	This release, in brief:
        - improves clarity through nomenclature:
              * renames "(output layer) backend implementation" to
                "(capture) strategy", for more standard nomenclature
                (and less of a mouthful);
              * renames "pseudo-layer" to "multi-layer" (if so
                declared) or "non-trapping layer" (neither is a direct
                analogue to PerlIO pseudo layers, so the nomenclature
                was misleading);
        - adds (import) options to the TempFile and SystemSafe capture
	  strategy factories, allowing for different ways to handle
	  PerlIO layers;
        - using these for two new standard capture strategies,
          "tempfile-preserve" and "systemsafe-preserve"; and
        - fixes a bug in SystemSafe.
        And, in more detail ...
        Test::Trap::Builder:
        - Changes method names per the nomenclature changes, leaving
          back-compat aliases behind.
        - Changes error message per the nomenclature changes.  (No
          back-compat possible, sorry.)
        - Updates the documentation.
        Test::Trap::Builder::{TempFile,SystemSafe}:
        - Import now takes arguments:
		strategy name (default {"tempfile","systemsafe"}); and
                strategy options (default empty hash).
        - The following options are supported:
		preserve_io_layers (boolean; default false); and
        	io_layers (colon-separated string; default unset).
        Test::Trap::Builder::SystemSafe:
        - Fixes a bug where the original perlio layers were not
	  restored after the trap was sprung.
        Test::Trap::Builder::{PerlIO,TempFile,SystemSafe}:
        - Updates the code in accordance with Test::Trap::Builder
          changes; see above.
        - Updates the documentation.
        Test::Trap:
        - Imports new capture strategy, "tempfile-preserve", from
          TempFile with option preserve_io_layers.
        - Imports new capture strategy, "systemsafe-preserve", from
          SystemSafe with option preserve_io_layers.
        - Updates the code in accordance with Test::Trap::Builder
          changes; see above.
        - Updates the documentation.
        Tests:
        - changes variable names and comments in accordance with
          nomenclature changes;
        - adds tests for the no-restore bug in t/03-files.pl (hence
          for each tested strategy);
        - adds t/03-files-{tempfile,systemsafe}-preserve.t to run the
          basic tests against the new capture strategies;
        - accounts for the changed error message in t/06-layers.t;
        - adds t/{15-tempfile,16-systemsafe}-options.t to check
          capture strategies of varying options.

0.2.5	Sun Nov 16 18:31:42 CET 2014
	This release localizes $! (ERRNO) for internal operations that
	change it, as suggested by Felipe Gasper.  For the same
	operations, it also localizes $^E (extended OS error).
        Library files:
        - Localize $! and $^E for internal operations that change them.
        Tests:
        - Check that bare test blocks leave $!, %!, and $^E unchanged,
          but that blocks that modify $! and $^E still do so.
        Also:
        - Added Module::Build to the configure_requires.
        - Fix emacs mode line on t/14-leaks.t.
        - Better comments in lib/Test/Trap/Builder.pm.

0.2.4	Sun Mar 30 10:02:16 CEST 2014
	This release fixes a memory leak with the default tempfile
	backend layer implementation for trapping output.  The
	systemsafe alternative backend had the same kind of leak, and
	has also been fixed.  (The perlio alternative backend was not
	affected.)
        Thanks go again to Felipe Gasper for discovering this!
        Test::Trap::Builder:
        - Add a method ExceptionFunction to make it easier to avoid
          circular references arising from registered closures.
          Document and export this method.
        Test::Trap::Builder::TempFile:
        - Break a circular reference chain.
        Test::Trap::Builder::SystemSafe:
        - Break a similar but slightly more complicated circular
          reference chain.
        Tests:
        - Added t/14-leaks.t to check that the reference count is as
          expected, at least with trivial code and standard setup for
          all three output layer backends.
        Also:
        - Use ebhanssen@cpan.org for my email address.
        - Update copyright years.
        - Update MANIFEST.SKIP &c for use of git (goodbye subversion).

0.2.3   Mon Dec 30 14:46:15 CET 2013
        Library files:
        - RT #87263: typo fixes, courtesy dsteinbrunner@pobox.com.
        - Localize the trap coderef, so that it does not hold on to
          what it closes over, creating stale references, but lets
          them go out of scope; thanks go to Felipe Gasper.
        Documentation:
        - Some clarification for RT #91687: Wrong exit code after die.
        Tests:
        - Regression test for the stale references.

0.2.2	Fri Mar  2 02:23:04 CET 2012
        Test::Trap::Builder::TempFile & Test::Trap::Builder::SystemSafe:
        - RT #61776: Remove temporary files at first opportunity;
          thanks go to Ian Goodacre.
        Documentation:
        - RT #61773: return from trap function
          Document what trap { ... } returns.
        Tests:
        - RT #75430: Failing tests in 06-layer.t
          Convert a number of die_is to die_like, handling cases with
          and without trailing comma from carp.
        - Unset another taint-checked environment variable, CDPATH;
          thanks go to CPAN testers.

0.2.1	Tue Dec 29 23:14:45 CET 2009
	Documentation:
	 - fix a typo, [RT #48941]; thanks go to David Taylor;
	Tests:
	 - use .PL-files to build the test files t/08-fork.t and
           t/11-systemsafe-basic.t instead of dispatching through
           *-taint.t and *-no-taint.t files at runtime;
	Build:

t/08-fork.pl
#!perl
# -*- mode: cperl ; compile-command: "cd .. ; ./Build ; prove -vb t/08-*.t" -*-
use Test::More tests => 15;
use strict;
use warnings;

my $flag;
BEGIN {
  *CORE::GLOBAL::exit = sub(;$) {
    if ($flag) {
      pass("The final test: The outer CORE::GLOBAL::exit is eventually called");
    }
    else {
      fail("The outer CORE::GLOBAL::exit is called too soon!");
    }
    CORE::exit(@_ ? shift : 0);
  };
}

BEGIN {
  use_ok( 'Test::Trap' );
}

# check that the setup works -- the exit is still trapped:
trap { exit };
is( $trap->exit, 0, "Trapped the first exit");

# check that the exit from the forked-off process reverts to the inner
# CORE::GLOBAL::exit, not the outer
trap {
  *CORE::GLOBAL::exit = sub(;$) {
    pass("The inner CORE::GLOBAL::exit is called from the child");
    CORE::exit(@_ ? shift : 0);
  };
  trap {
    fork;
    exit;
  };
  wait; # let the child finish first
  # Increment the counter correctly ...
  my $Test = Test::More->builder;
  $Test->current_test( $Test->current_test + 1 );
  is( $trap->exit, 0, "Trapped the inner exit");
};
like( $trap->stderr, qr/^Subroutine (?:CORE::GLOBAL::)?exit \Qredefined at ${\__FILE__} line/, 'Override warning' );

trap {
  trap{
    trap {
      fork;
      exit;
    };
    wait;
    is( $trap->exit, 0, "Trapped the inner exit" );
  }
};
is( $trap->leaveby, 'return', 'Should return just once, okay?' );

# Output from forked-off processes?
my $me;
trap {
  $me = fork ? 'parent' : 'child';
  print "\u$me print\n";
  warn "\u$me warning\n";
  wait, exit $$ if $me eq 'parent';
};
CORE::exit(0) if $me eq 'child';
is( $trap->exit, $$, "Trapped the parent exit" );
like( $trap->stdout, qr/^(Parent print\nChild print\n|Child print\nParent print\n)/, 'STDOUT from both processes!' );
like( $trap->stderr, qr/^(Parent warning\nChild warning\n|Child warning\nParent warning\n)/, 'STDERR from both processes!' );
is_deeply( $trap->warn, ["Parent warning\n"], 'Warnings from the parent only' );

# STDERR from forked-off processes, with a closed STDIN & STDOUT?
trap {
  close STDOUT;
  trap {
    my $me = fork ? 'parent' : 'child';
    print "\u$me print\n";
    warn "\u$me warning\n";
    wait, exit $$ if $me eq 'parent';
    CORE::exit(0);
  };
  is( $trap->exit, $$, "Trapped the parent exit" );
  is( $trap->stdout, '', 'STDOUT from both processes is nil -- the handle is closed!' );
  like( $trap->stderr, qr/\A(?=.*^Parent warning$)(?=.*^Child warning$)/ms, 'STDERR from both processes!' );
};

$flag++; # the exit test will now pass -- in the forked-off processes it will fail!
exit;


t/03-files.pl
#!perl -T

BEGIN { $_ = defined && /(.*)/ && $1 for @ENV{qw/ TMPDIR TEMP TMP /} } # taint vs tempfile
use Test::More;
use IO::Handle;
use File::Temp qw( tempfile );
use Data::Dump qw(dump);
use strict;
use warnings;

our $strategy; # to be set in the requiring test script ...
BEGIN {
  my $pkg = "Test::Trap::Builder::$strategy";
  local $@;
  eval qq{ use $pkg };
  if (exists &{"$pkg\::import"}) {
    plan tests => 1 + 6*10 + 5*3 + 17; # 10 runtests; 3 inner_tests; another bunch ...
  }
  else {
    plan skip_all => "$strategy strategy not supported; skipping";
  }
}

# This is an ugly bunch of tests, but for regression's sake, I'll
# leave it as-is.

# One problem is that warn() (or rather, the default __WARN__ handler)
# will print on the previous STDERR if the current STDERR is closed.

# Another problem is that the __WARN__ handler has not always been
# properly restored on exit from a trap.  Ouch.

BEGIN {
  use_ok( 'Test::Trap', '$T', lc ":flow:stdout($strategy):stderr($strategy):warn" );
}

STDERR: {
  close STDERR;
  my ($errfh, $errname) = tempfile( UNLINK => 1 );
  open STDERR, '>', $errname;
  STDERR->autoflush(1);
  print STDOUT '';
  sub stderr () { local $/; no warnings 'io'; local *ERR; open ERR, '<', $errname or die; <ERR> }
  END { close STDERR; close $errfh }
}

sub diagdie {
  my $msg = shift;
  diag $msg;
  die $msg;
}

my ($noise, $noisecounter) = ('', 0);
sub runtests(&@) { # runs the trap and performs 6 tests
  my($code, $return, $warn, $stdout, $stderr, $desc) = @_;
  my $n = ++$noisecounter . $/;
  warn $n or diagdie "Cannot warn()!";
  STDERR->flush or diagdie "Cannot flush STDERR!";
  print STDERR $n or diagdie "Cannot print on STDERR!";
  STDERR->flush or diagdie "Cannot flush STDERR!";
  $noise .= "$n$n";
  $warn = do { local $" = "[^`]*`"; qr/\A@$warn[^`]*\z/ };
  my @r = eval { &trap($code) }; # bypass prototype
  my $e = $@;
SKIP: {
    ok( !$e, "$desc: No internal exception" ) or do {
      diag "Got internal exception: '$e'";
      skip "$desc: Internal exception -- bad state", 5;
    };
    is_deeply( $T->return, $return, "$desc: Return" );
    like( join("`", @{$T->warn}), $warn, "$desc: Warnings" );
    is( $T->stdout, $stdout, "$desc: STDOUT" );
    like( $T->stderr, $stderr, "$desc: STDERR" );
    is( stderr, $noise, ' -- no uncaptured STDERR -- ' );
  }
}

my $inner_trap;
sub inner_tests(@) { # performs 5 tests
  my($return, $warn, $stdout, $stderr, $desc) = @_;
  $warn = do { local $" = "[^`]*`"; qr/\A@$warn[^`]*\z/ };
SKIP: {
    ok(eval{$inner_trap->isa('Test::Trap')}, "$desc: The object" )
      or skip 'No inner trap object!', 4;
    is_deeply( $inner_trap->return, $return, "$desc: Return" );
    like( join("`", @{$inner_trap->warn}), $warn, "$desc: Warnings" );
    is( $inner_trap->stdout, $stdout, "$desc: STDOUT" );
    like( $inner_trap->stderr, $stderr, "$desc: STDERR" );
  }
  undef $inner_trap; # catch those simple mistakes.
}

runtests { 5 }
  [5], [],
  '', qr/\A\z/,
  'No output';

runtests { my $t; print "Test printing '$t'"; 2}
  [2], [ qr/^Use of uninitialized value.* in concatenation \Q(.) or string at / ],
  "Test printing ''", qr/^Use of uninitialized value.* in concatenation \Q(.) or string at /,
  'Warning';

runtests { close STDERR; my $t; print "Test printing '$t'"; 2}
  [2], [ qr/^Use of uninitialized value.* in concatenation \Q(.) or string at / ],
  "Test printing ''", qr/\A\z/,
  'Warning with closed STDERR';

runtests { warn "Testing stderr trapping\n"; 5 }
  [5], [ qr/^Testing stderr trapping$/ ],
  '', qr/^Testing stderr trapping$/,
  'warn()';

runtests { close STDERR; warn "Testing stderr trapping\n"; 5 }
  [5], [ qr/^Testing stderr trapping$/ ],
  '', qr/\A\z/,
  'warn() with closed STDERR';

runtests {
  warn "Outer 1st\n";
  my @r = trap { warn "Testing stderr trapping\n"; 5 };
  binmode(STDERR); # XXX: masks a real weakness -- we do not simply restore the original!
  $inner_trap = $T;
  warn "Outer 2nd\n";
  @r
} [5], [ qr/Outer 1st/, qr/Outer 2nd/ ],
  '', qr/^Outer 1st\nOuter 2nd$/,
  'warn() in both traps';
inner_tests
  [5], [ qr/^Testing stderr trapping$/ ],
  '', qr/^Testing stderr trapping$/,
  ' -- the inner trap -- warn()';

runtests { print STDERR "Test printing"; 2}
  [2], [],
  '', qr/^Test printing\z/,
  'print() on STDERR';

runtests { close STDOUT; print "Testing stdout trapping\n"; 6 }
  [6], [ qr/^print\Q() on closed filehandle STDOUT at / ],
  '', qr/^print\Q() on closed filehandle STDOUT at /,
  'print() with closed STDOUT';

runtests { close STDOUT; my @r = trap { print "Testing stdout trapping\n"; (5,6) }; $inner_trap = $T; @r }
  [5, 6], [],
  '', qr/\A\z/,
  'print() in inner trap with closed STDOUT';
inner_tests
  [5, 6], [ qr/^print\Q() on closed filehandle STDOUT at / ],
  '', qr/^print\Q() on closed filehandle STDOUT at /,
  ' -- the inner trap -- print() with closed STDOUT';

runtests { close STDERR; my @r = trap { warn "Testing stderr trapping\n"; 2 }; $inner_trap = $T; @r }
  [2], [],
  '', qr/\A\z/,
  'warn() in inner trap with closed STDERR';
inner_tests
  [2], [ qr/^Testing stderr trapping$/ ],
  '', qr/\A\z/,
  ' -- the inner trap -- warn() with closed STDERR';

# regression test for the ', <$fh> line 1.' bug:
trap {
    trap {};
    warn "no newline";
};
unlike $T->stderr, qr/, \S+ line 1\./, 'No "<$f> line ..." stuff, please';

# regression test for preservation of PerlIO layers:
SKIP: {
  skip 'Lacking PerlIO', 4 unless eval "use PerlIO; 1";
  my @io = PerlIO::get_layers(*STDOUT);
  trap { binmode STDOUT, ':utf8' }; # or whatever, really
  is_deeply( [PerlIO::get_layers(*STDOUT)], \@io, 'STDOUT still has the original layers.')
    or diag(dump(\@io));
  binmode STDOUT;
  my @raw = PerlIO::get_layers(*STDOUT);
  trap { binmode STDOUT, ':utf8' }; # or whatever, really
  is_deeply( [PerlIO::get_layers(*STDOUT)], \@raw, 'STDOUT is still binmoded.')
    or diag(dump([PerlIO::get_layers(*STDOUT)], \@raw));
  binmode STDOUT, ':crlf';
  my @crlf = PerlIO::get_layers(*STDOUT);
  trap { binmode STDOUT, ':utf8' }; # or whatever, really
  is_deeply( [PerlIO::get_layers(*STDOUT)], \@crlf, 'STDOUT still has the crlf layer(s).')
    or diag(dump([PerlIO::get_layers(*STDOUT)], \@crlf));
  binmode STDOUT;
  my @tmp = @io;
  $_ eq $tmp[0] ? shift @tmp : last for PerlIO::get_layers(*STDOUT);
  binmode STDOUT, $_ for @tmp;
  is_deeply( [PerlIO::get_layers(*STDOUT)], \@io, 'Sanity check: STDOUT now again has the original layers.')
    or diag(dump([PerlIO::get_layers(*STDOUT)], \@io));
}

# test the $! handling:
my $errnum = 11; # "Resource temporarily unavailable" locally -- sounds good :-P
my $errstring = do { local $! = $errnum; "$!" };
my $erros = do { local $! = $errnum; $^E };
my ($errsym) = do { local $! = $errnum; grep { $!{$_} } keys(%!) };
for my $case ([Bare => sub { return 42 }], [Dying => sub { die 42 }], [Exiting => sub { exit 42 }]) {
  local $! = $errnum;
  trap {};
  my ($sym) = grep { $!{$_} } keys(%!);
  is $!+0, $errnum, "$strategy trap doesn't change errno (remains $errnum/$errstring)";
  is $^E,  $erros,  "$strategy trap doesn't change extended OS error (remains $erros)";
  is $sym, $errsym, "$strategy trap doesn't change the error symbol (remains $errsym)";
}

{
  local $! = $errnum;
  trap {
    $! = 0;
    $^E = '';
  };
  my ($sym) = grep { $!{$_} } keys(%!);
  is $!+0,     0, "Errno-unsetting trap unsets errno (it's not localized)";
  is $^E,     '', "Errno-unsetting trap unsets extended OS error (it's not localized)";
  is $sym, undef, "Errno-unsetting trap unsets the error symbol (it's not localized)";
}

1;


Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.4003
require 5.006002;
use ExtUtils::MakeMaker;
WriteMakefile
(
  'NAME' => 'Test::Trap',
  'VERSION_FROM' => 'lib/Test/Trap.pm',
  'PREREQ_PM' => {
                   'Carp' => 0,
                   'Data::Dump' => 0,
                   'Exporter' => 0,
                   'File::Temp' => 0,
                   'IO::Handle' => 0,
                   'Test::Builder' => 0,
                   'Test::More' => 0,
                   'Test::Tester' => '0.107',
                   'base' => 0,
                   'constant' => 0,
                   'lib' => 0,
                   'strict' => 0,
                   'version' => 0,
                   'warnings' => 0
                 },
  'INSTALLDIRS' => 'site',
  'EXE_FILES' => [],
  'PL_FILES' => {
                  't/08-fork.PL' => 't/08-fork.t',
                  't/11-systemsafe-basic.PL' => 't/11-systemsafe-basic.t'
                }
)
;


Build.PL
use strict;
use warnings;
use Module::Build;

my $class = Module::Build->subclass( code => do { local $/; <DATA> } );

my $builder = $class->new
  ( module_name         => 'Test::Trap',
    license             => 'perl',
    create_makefile_pl  => 'traditional',
    dist_author         => 'Eirik Berg Hanssen <ebhanssen@cpan.org>',
    dist_version_from   => 'lib/Test/Trap.pm',
    requires =>
    { 'base'              => 0, # core ...
      'constant'          => 0, # core ...
      'Carp'              => 0,
      'Data::Dump'        => 0,
      'Exporter'          => 0,
      'File::Temp'        => 0,
      'IO::Handle'        => 0,
      'lib'               => 0, # core ...
      'strict'            => 0, # now _that's_ core!
      'Test::Builder'     => 0,
      'Test::More'        => 0,
      'Test::Tester'      => 0.107,   # fails with at least some earlier ones
      'perl'              => '5.6.2',
      'version'           => 0,
      'warnings'          => 0,
    },
    PL_files =>
    { 't/08-fork.PL'             => 't/08-fork.t',
      't/11-systemsafe-basic.PL' => 't/11-systemsafe-basic.t',
    },
    add_to_cleanup =>
    [ 'Test-Trap-*',
      'MYMETA.yml',
      't/08-fork.t',
      't/11-systemsafe-basic.t',
    ],
  );

$builder->create_build_script();

__DATA__

=head1 ACTIONS

=head2 authortest

This runs all the C<xt/author> tests, as well as the ordinary tests,
after making sure that the build, manifest, and distmeta actions have
been taken.

=cut

sub ACTION_authortest {
  my ($self) = @_;

  $self->depends_on('build');
  $self->depends_on('manifest');
  $self->depends_on('distmeta');

  $self->test_files( qw< t xt/author > );
  $self->recursive_test_files(1);

  $self->depends_on('test');

  return;
}

sub ACTION_distdir {
  my ($self) = @_;

  $self->depends_on('authortest');

  return $self->SUPER::ACTION_distdir();
}

sub ACTION_distmeta {
  my ($self) = @_;

  require Module::Build::Compat;
  unless (Module::Build::Compat->VERSION gt 0.31 or eval { Module::Build::Compat->PL_FILES_PATCH }) {
    die <<'DIE'
Too old Module::Build::Compat to Build distmeta.
Upgrade if possible or apply the following hack patch:
--- Compat.pm   2008-10-04 02:14:02.000000000 +0200
+++ Compat.pm   2008-10-04 02:15:10.000000000 +0200
@@ -139,7 +139,7 @@
     
     $MM_Args{EXE_FILES} = [ sort keys %{$build->script_files} ] if $build->script_files;
     
-    $MM_Args{PL_FILES} = {};
+    $MM_Args{PL_FILES} = $build->PL_files;
     
     local $Data::Dumper::Terse = 1;
     my $args = Data::Dumper::Dumper(\%MM_Args);
@@ -153,6 +153,7 @@
   }
 }
 
+sub PL_FILES_PATCH { 1 }
 
 sub subclass_dir {
   my ($self, $build) = @_;
DIE
  }

  return $self->SUPER::ACTION_distmeta();
}
